### ISSUE TITLE
AgentError: Fetch implementation was not available. You appear to be in a Node.js context, but global.fetch was not available.

### ISSUE THREAD

**Issue by tokuryoo** [canscale/hello-candb#1]
I couldn't find a Discord or something, so I created this Issue. I have worked according to the following.
`git clone -b beta git@github.com:canscale/hello-candb.git`
https://docs.google.com/document/d/1lEzMtAnM7ac8dE8zLZ_SE7a5jxgcZpC_XeUz2SK3WQ8/edit

`npm run upgrade` produced the following error.
>AgentError: Fetch implementation was not available. You appear to be in a Node.js context, but global.fetch was not available.
>     at getDefaultFetch (/Users/tokuryo/dfinity/github/hello-candb/extended_examples/hello_world_with_upgrades/management_scripts/node_modules/@dfinity/agent/src/agent/http/index.ts:119:13)
>     at new HttpAgent (/Users/tokuryo/dfinity/github/hello-candb/extended_examples/hello_world_with_upgrades/management_scripts/node_modules/@dfinity/agent/src/agent/http/index.ts:166:38)

Did I do something wrong? Or is it a bug?

dfx 0.14.0
macOS Monterey 12.6.5

**Comment by junkei-okinawa** [comment]
I also got the same Error.
I think the error occurs in `initializeIndexClient` used in `upgrade.tsx`. 
I think `new IndexClient<IndexCanister>` can only be used from the Browser. Am I wrong? 
I implemented the following as a trial and it succeeded in upgrading helloservice canister. 
1. port the function of `upgrade.ts` to frontend.
  1-1. put the wasm loading part to `webpack.config.js`.
  1-2. port actor call part to `src/App.tsx` by making function.
2. make Button element in `src/App.tsx` and call function 1-2 by Onclick.

```webpack.config.js
+ const { loadWasm } = require("candb-client-typescript/dist/ClientUtil");
+ const helloServiceWasmModulePath = `${process.cwd()}/../.dfx/local/canisters/helloservice/helloservice.wasm`
+ prev["HELLO_SERVICE_WASM"] = loadWasm(helloServiceWasmModulePath);
```

```src/App.tsx
  async function upgradeIndexCanister() {
    const upgradeResult = await indexClient.indexCanisterActor.upgradeGroupCanistersInPKRange("group#", "group#~", process.env.HELLO_SERVICE_WASM);
    console.log("result", JSON.stringify(upgradeResult));
    console.log("upgrade complete");
  }

      <div className="section-wrapper">
        <h2>Upgrade index canister</h2>
        <div className="flex-wrapper">
          <button type="button" onClick={() => upgradeIndexCanister()}>upgrade</button>
        </div>
      </div>
```

**Comment by tokuryoo** [comment]
@junkei-okinawa 
Thank you for your reply.

hello-candb/extended_examples/hello_world_with_upgrades/dfx.json
>  "declarations": {
>     "output": "src/declarations",
>     "node_compatibility": "true"
>   },

Maybe "node_compatibility": "true" is not working. I did not know why it did not work.

**Comment by junkei-okinawa** [comment]
It worked with below fix.

dfx.json
before
```dfx.json
{
  "version": 1,
  "canisters": {
    "index": {
      "type": "motoko",
      "main": "src/index/IndexCanister.mo"
    },
    "helloservice": {
      "type": "motoko",
      "main": "src/helloservice/HelloService.mo"
    }
  },
  "declarations": {
    "output": "src/declarations",
    "node_compatibility": "true"
  },
  "defaults": {
    "build": {
      "packtool": "vessel sources",
      "args": ""
    }
  },
  "networks": {
    "local": {
      "bind": "127.0.0.1:8000",
      "type": "ephemeral"
    }
  }
}
```

after
```dfx.json
{
  "version": 1,
  "canisters": {
    "index": {
      "type": "motoko",
      "main": "src/index/IndexCanister.mo",
+      "declarations": {
+        "node_compatibility": true
      }
    },
    "helloservice": {
      "type": "motoko",
      "main": "src/helloservice/HelloService.mo",
+      "declarations": {
+        "node_compatibility": true
      }
    }
  },
  "declarations": {
    "output": "src/declarations"
-   "node_compatibility": "true"
  },
  "defaults": {
    "build": {
      "packtool": "vessel sources",
      "args": ""
    }
  },
  "networks": {
    "local": {
      "bind": "127.0.0.1:8000",
      "type": "ephemeral"
    }
  }
}
```

**Comment by junkei-okinawa** [comment]
Created a PullRequest for the fix
https://github.com/canscale/hello-candb/pull/2

**Comment by ByronBecker** [comment]
Thanks for this fix! What version of Node are you using? I've used Node 16 in the past to get around this issue. I think the node compatibility is a much better solution.

Btw, merging in your fix! Thanks for the contribution.

**Comment by ByronBecker** [comment]
Fixed in #2 & https://github.com/canscale/hello-candb/pull/3

**Comment by junkei-okinawa** [comment]
@ByronBecker 
> What version of Node are you using?

I also using Node16.
